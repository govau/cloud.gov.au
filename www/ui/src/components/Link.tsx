import * as React from "react";
import { Link as RouterLink } from "react-router-dom";
import styled, { css } from "../styled-components";

import Theme from "../theme";
import { LinkProps as ReactRouterLinkProps } from "react-router-dom";

export interface BaseLinkProps {
  className?: string;
}

export interface RouterLinkProps extends BaseLinkProps, ReactRouterLinkProps {}

export interface LinkProps extends BaseLinkProps {
  href?: string;
}

export const isRouterLink = (
  link: Partial<RouterLinkProps>
): link is RouterLinkProps => link.to !== undefined;

const styles = ({ theme }: { theme: Theme }) =>
  css`
    color: ${theme.linkColor};
    border-bottom: solid 1px ${theme.linkUnderlineColor};
    &:hover {
      color: ${theme.linkHoverColor};
      background-color: ${theme.linkHoverBgColor};
      border-bottom-color: ${theme.linkHoverBgColor};
    }
  `;

const StyledLink = styled.a`
  transition: 0.25s;
  text-decoration: none;
  ${styles};
`;

const StyledRouterLink = styled(RouterLink)`
  transition: 0.25s;
  text-decoration: none;
  ${styles};
`;

const Link: React.StatelessComponent<LinkProps | RouterLinkProps> = props => {
  const { children, className } = props;
  if (isRouterLink(props)) {
    const { to, ...rest } = props;
    return (
      <StyledRouterLink className={className} to={to} {...rest}>
        {children}
      </StyledRouterLink>
    );
  }
  const { href } = props;
  return (
    <StyledLink className={className} href={href}>
      {children}
    </StyledLink>
  );
};

export default Link;
