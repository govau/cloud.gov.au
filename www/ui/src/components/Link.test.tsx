import * as React from "react";
import { MemoryRouter } from "react-router";
import * as renderer from "react-test-renderer";

import Link from "./Link";

it("renders without crashing", () => {
  const tree = renderer
    .create(
      <MemoryRouter>
        <Link to="/">Home</Link>
      </MemoryRouter>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
