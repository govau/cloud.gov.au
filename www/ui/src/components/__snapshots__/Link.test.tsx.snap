// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders without crashing 1`] = `
<a
  className="sc-bwzfXH fZcoTc"
  href="/"
  onClick={[Function]}
>
  Home
</a>
`;
